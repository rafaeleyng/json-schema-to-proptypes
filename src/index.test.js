import jsonSchemaToPropTypes, { SchemaSymbol } from './'

describe('jsonSchemaToPropTypes', function() {
  let schema, validators

  beforeEach(function() {
    schema = {
      "type" : "object",
      "description" : "PropTypes for some component",
      "properties" : {
        "id" : {
          "type" : "string"
        }
      }
    }

    validators = jsonSchemaToPropTypes(schema)
  })

  it('throws if json schema is not an object', function() {
    const schema = ""
    expect(() => jsonSchemaToPropTypes(schema)).toThrowErrorMatchingSnapshot()
  })

  it ('throws if json schema\'s type is not "object"', function() {
    const schema = {
      type: "string"
    }

    expect(() => jsonSchemaToPropTypes(schema)).toThrowErrorMatchingSnapshot()
  })

  it('creates a proptype validator for each prop', function() {
    Object.keys(schema.properties).forEach(function(propKey) {
      expect(validators[propKey]).toBeDefined()
      expect(typeof validators[propKey]).toBe('function')
    })
  })

  it('exposes the schema that it is using to validate', function() {
    const propTypes = jsonSchemaToPropTypes(schema)
    expect(propTypes[SchemaSymbol]).toEqual(schema)
  })

  it('hides the schema from enumerable properties', function() {
    const propTypes = jsonSchemaToPropTypes(schema)
    expect(Object.keys(propTypes)).toEqual(["id"])
  })

  it('can take existing proptypes as an argument', function() {
    const newSchema = {
      [SchemaSymbol]: schema
    }

    const propTypes = jsonSchemaToPropTypes(newSchema)
    expect(Object.keys(propTypes)).toEqual(["id"])
  })

  it('deep merges multiple schemas together', function() {
    const updates = {
      "properties" : {
        "name" : {
          "type" : "string"
        }
      }
    }

    const propTypes = jsonSchemaToPropTypes(schema, updates)
    expect(Object.keys(propTypes)).toEqual(["id", "name"])
  })

  describe('validator', function() {
    it('validates correctly based on the property it references', function() {
      const propTypes = jsonSchemaToPropTypes(schema)
      expect(propTypes["id"]({ id: 15 }, "id")).toMatchSnapshot()
      expect(propTypes["id"]({ id: "hello" }, "id")).toMatchSnapshot()
    })

    it('respects required properties', function() {
      const propTypes = jsonSchemaToPropTypes({
        type: "object",
        required: ["a"],
        properties: {
          a: { type: "string" },
          b: { type: "string" }
        }
      })

      expect(propTypes.a({}, "a")).toMatchSnapshot()
      expect(propTypes.b({}, "b")).toMatchSnapshot()
    })

    it('error on deep (nested) validation errors', function() {
      const propTypes = jsonSchemaToPropTypes({
        type: 'object',
        properties: {
          a: {
            type: 'object',
            properties: {
              b: { type: 'string' }
            }
          },
          c: { type: 'number' }
        }
      })
      expect(propTypes.a({a: {b: 1}, c: 1}, 'a')).toMatchSnapshot()
      expect(propTypes.c({a: {b: '1'}, c: 'dsasd'}, 'c')).toMatchSnapshot()
    })
  })
})
