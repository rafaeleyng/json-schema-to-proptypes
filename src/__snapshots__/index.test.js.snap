// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsonSchemaToPropTypes throws if json schema is not an object 1`] = `"Schema must be of type 'object'"`;

exports[`jsonSchemaToPropTypes throws if json schema's type is not "object" 1`] = `"Schema must define an object type (currently: string)"`;

exports[`jsonSchemaToPropTypes validator error on deep (nested) validation errors 1`] = `[Error: '.a.b' should be string, found {"b":1} instead. Check propTypes of component undefined]`;

exports[`jsonSchemaToPropTypes validator error on deep (nested) validation errors 2`] = `[Error: '.c' should be number, found "dsasd" instead. Check propTypes of component undefined]`;

exports[`jsonSchemaToPropTypes validator respects required properties 1`] = `[Error: '' should have required property 'a', found undefined instead. Check propTypes of component undefined]`;

exports[`jsonSchemaToPropTypes validator respects required properties 2`] = `null`;

exports[`jsonSchemaToPropTypes validator validates correctly based on the property it references 1`] = `[Error: '.id' should be string, found 15 instead. Check propTypes of component undefined]`;

exports[`jsonSchemaToPropTypes validator validates correctly based on the property it references 2`] = `null`;
